### PR TITLE
fix(windows): fix menu popup position

### LINKED ISSUE #4013: feat(windows): Keyman menu pops up to left of tray icon; it should probably center over icon for better usability
**Is your feature request related to a problem? Please describe.**

Keyman menu pops up to left of tray icon; it should probably center over icon for better usability

When implementing this, consider other locations of taskbar apart from bottom of screen!

---

**Keyman for Windows/macOS/Linux:**
 - OS: Windows 10 x64 1909 
 - Keyman version: 14.0.192-alpha

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #4013.

This turned out to be a lot more complicated than I expected, mostly due to limitations in the Windows Shell API with regard to multiple monitors and per-monitor DPI awareness (which we don't yet do in Keyman -- a task for another day). Some of the Shell API functions are also re-entrant, which is dangerous, so I avoided using those.

We now popup the menu relative to the Keyman icon far more precisely which gives a much nicer feel to the menu interaction.

This also fixes the menu popping up on the screen where the visual keyboard is visible, rather than the screen next to the icon.

Apologies for the number of screenshots below, but there were a lot of permutations to test. I tested on my Macbook with High DPI (2560x1600 @ 175%) on the laptop screen, and dual 1920x1080 external screens.

Because we currently lack DPI awareness in Keyman, the Keyman menu will be fuzzy on higher resolution screens. That's a separate project for a future version.

bottom-hiddenicon
bottom-visibleicon
hidpi-bottom-hiddenicon
hidpi-bottom-visibleicon
hidpi-left-hiddenicon
hidpi-left-visibleicon
hidpi-right-hiddenicon
hidpi-right-visibleicon
hidpi-top-hiddenicon
hidpi-top-visibleicon
left-hiddenicon
left-visibleicon
right-hiddenicon
right-visibleicon
top-hiddenicon
top-visibleicon

### CODE REVIEW

File: windows/src/engine/keyman/UfrmKeymanMenu.dfm [hunk 1]
```diff
@@ -6,6 +6,7 @@ object frmKeymanMenu: TfrmKeymanMenu
   ClientHeight = 632
   ClientWidth = 707
   Color = clBtnFace
+  DefaultMonitor = dmDesktop
```
Comment (mcdurdin) [inline windows/src/engine/keyman/UfrmKeymanMenu.dfm:9]: This prevents the menu popping on the same monitor as the osk

File: windows/src/engine/keyman/UfrmKeymanMenu.pas [hunk 2]
```diff
@@ -681,13 +686,13 @@ procedure TfrmKeymanMenu.DrawLayer;   // I3990
...
 
-procedure TfrmKeymanMenu.FitToDesktop;
+procedure TfrmKeymanMenu.FitToDesktop(X, Y: Integer);
 var
   m: TMonitor;
   wr, mr: TRect;
 begin
   wr := Rect(Left,Top,Left+Width,Top+Height);
-  m := Screen.MonitorFromRect(wr, mdNearest);
+  m := Screen.MonitorFromPoint(Point(X, Y), mdNearest);
```
Comment (mcdurdin) [inline windows/src/engine/keyman/UfrmKeymanMenu.pas:695]: It is possible the workarea has other taskbars which impact the calculation of how to fit the menu onto the screen. However I have not encountered this in testing at this time.

File: windows/src/engine/keyman/UfrmKeyman7Main.pas [hunk 3]
```diff
@@ -983,60 +983,136 @@ procedure TfrmKeyman7Main.LanguageSwitchFormHidden(Sender: TObject);
...
+      end;
+    end;
+
+    // Align the menu according to the location of the taskbar
+    case systraylocation of
+      alTop:    begin pt.X := (r.Right + r.Left) div 2; pt.Y := r.Bottom; end;
+      alBottom: begin pt.X := (r.Right + r.Left) div 2; pt.Y := r.Top; end;
+      alLeft:   begin pt.X := r.Right; pt.Y := r.Top; end;
+      alRight:  begin pt.X := r.Left; pt.Y := r.Top; end;
+    end;
```
Comment (jahorton) [inline windows/src/engine/keyman/UfrmKeyman7Main.pas:1083]: This seems pretty close to windows/src/engine/keyman/UfrmKeymanMenu.pas, line 478+.  Any reason for the WETness here?

Comment (mcdurdin) [reply windows/src/engine/keyman/UfrmKeyman7Main.pas:1083]: This here calculates the anchor point of the menu; the code in UfrmKeymanMenu calculates the actual position of the menu with reference to the anchor point, once we know its width and height.

We could move these changes into the UfrmKeymanMenu unit but the calculations are different in each case so it's not going to change much.

**Review by jahorton — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 14.0.221-beta](https://keyman.com/downloads/releases/beta/14.0.221)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.9-alpha](https://keyman.com/downloads/releases/alpha/15.0.9)

**Pull request merged**